### ISSUE TITLE
higher sample rate

### ISSUE THREAD

**Issue by ChironeX1976** [waveshare/High-Pricision_AD_HAT#15]
Hello,
i was wondering how i can get a higher sample rate when reading a sensor. I use a Raspberry Pi with 8 gig RAM. CPU is set at 1200 Mhz. I am using the ADC-device to capture sound, so i need at least 20.000 samples per second.
When i run the by waveshare provided Main.py - TEST_ADC1_RATE at **38400**SPS, the output is:

        - ID Read success 
        - REG_MODE2 success
        - REG_REFMUX success
        - REG_MODE0 success
        - REG_MODE1 success
        - SPS-setting: ADS1263_38400SPS
        - ...start and endtime= 1690130018.1033027 1690130018.8256269
        - ...elapsed time for 5000 samples= 0.7223241329193115
        - SPS-result singlechannelfrequency =  6922.100165464822

As you can see, i want 38400 samples per second and i get approx. **7000** samples per second. 

Attached, also a plot of 10 seconds of audiosignal-recording. 10 seconds * 38400 SPS should be 384.000 points of data. The horizontal axis shows only 80.000 points.

What am i doing wrong?

Thanks for any advice.

Kris
![Figure_1](https://github.com/waveshare/High-Pricision_AD_HAT/assets/55454779/a41d194c-97a8-489e-b65a-2d5742be44f3)

**Comment by mahdiaslanimk** [comment]
Roughly the same problem here. @ChironeX1976 did you found any solutions?

**Comment by ChironeX1976** [comment]
No, i did not find a solution. The program in C is much faster than Python,
someone told me, but i did not do any experiments with C.
cheerz
Kris

Op ma 18 dec 2023 om 06:53 schreef Mahdi Aslani ***@***.***>:

**Comment by mahdiaslanimk** [comment]
@ChironeX1976 Read the FAQ in this [link](https://www.waveshare.com/wiki/High-Precision_AD_HAT). There you can find the reason of low sps and an [example code](https://files.waveshare.com/upload/9/95/38K_High-Precision_AD_HAT_Code.zip) with higher rate (in both c an python). Hope this helps.

**Comment by E-wallace0001** [comment]
I have just had a look at the c implementation on the pi zero2 W

Firstly, there are some optimizations you can do for the SPI communications; You can request multiple bytes from the lower level librarys, instead of going back and forth asking for each byte at a time. This will gain you approx 1KHz straight off.

Secondly, the clock is set to 1MHz, changing this will get you a much higher sampling rate. The clocks are hardcoded. BCM driver needs to set a clock divider, while the wiringpi, and dev lib need to be set. My clock seems to top out at 22MHz. This can be configured in /lib/config/dev_config.c under DEV_module_init.

With the above changes, 35KHz can be achieved with wiring pi.

I was a little bit surprised that not requesting the status bytes actually slowed down the sampling rate.

I feel a better interface to the underlying libraries would be beneficial.

Hope this helps.

**Comment by brendanwallison** [comment]
Hello, I was looking at buying this ADC for audio as well. I have a very basic question about the 38400 SPS. If you plug in multiple microphones, would the sample rate be divided by the number of microphones, or is this a per-channel figure?

**Comment by schwebke** [comment]
on my Raspberry Pi Zero 2 W I was able to speed up the python solution from 6.275 SPS to 20.500 SPS -- in case someone wants to give it try for more testing:
https://github.com/schwebke/High-Pricision_AD_HAT

**Comment by bibhu223** [comment]
Did anyone use Pi 5 for this ADC ? I am not able to run this with C as well as python .. because it doesn't have bcm2835 ...can anyone help me.

**Comment by mahdiaslanimk** [comment]
@bibhu223 to use the python library for RPi 5 do the following:
- install ```rpi-lgpio``` library
```bash
pip install rpi-lgpio
```
- delete the implementation condition at the end of the ```config.py``` file and force it to use RPi:
```python
#if hostname == "raspberrypi":
#    implementation = RaspberryPi()
#else:
#    implementation = JetsonNano()

implementation = RaspberryPi()
```
- run ```main.py``` (first cd to the folder containing this file)
```bash
Python3 main.py
```

**Comment by bibhu223** [comment]
![IMG20241001084849](https://github.com/user-attachments/assets/add3bf03-1cef-4a74-a52e-81c9edae0706)
I have already tried this way ,, but it's showing error 
Runtime error : Cannot determine SOC peripheral bas address

**Comment by bibhu223** [comment]
@mahdiaslanimk  Thank you ,, i reboot the system now it's working on python ,, but it's very slow SPS , I was using pi zero W , In C the speed was very good .. but in this case 4kHz in single channel maximum .. How to increase the speed , One more thing the sample rate is not consistent , in my python program I am using some software smoothing and filtering ,which requires constant sampling rate ..but during data acquisition the sample is fluctuating...how to deal with this .

**Comment by bibhu223** [comment]
@ChironeX1976 , Have you tried with raspberry pi 5 with C , I am not able to run on C with pi 5 .

**Comment by mahdiaslanimk** [comment]
@bibhu223 have you enabled SPI interface in the config:
```sudo raspi-config```

**Comment by bibhu223** [comment]
@mahdiaslanimk 
n@raspberrypi:~/adc/hat/c $ sudo ./main
ADS1263 Demo
/***********************************/
Current environment: Debian
Please make JETSON !!!!!!!!!!
Segmentation fault
...This is error I am getting

**Comment by mahdiaslanimk** [comment]
@bibhu223 As I said you have to force the code to use the RPi for the implementation. Next to the ```main.py``` file there is a ```config.py``` file. Edit last lines in this code to achieve what I said:
```python
#if hostname == "raspberrypi":
#    implementation = RaspberryPi()
#else:
#    implementation = JetsonNano()

implementation = RaspberryPi()
```
Save the changes and rerun ```main.py```.

**Comment by bibhu223** [comment]
@mahdiaslanimk  yes it's running on python. 
But the above issue is coming when I run in C . As I need little bit higher data rate ....so I was trying with C . But not able to run.

**Comment by mahdiaslanimk** [comment]
@bibhu223 i didn’t try the C codes. Have you tried @E-wallace0001 ‘s suggestions or @schwebke ’s codes to speed up data acquisition using python?

**Comment by bibhu223** [comment]
@mahdiaslanimk yes , but still if I am plotting the voltage data in real time ...the frequency of data rate is only 40-60Hz ,,

**Comment by E-wallace0001** [comment]
> [@mahdiaslanimk](https://github.com/mahdiaslanimk) n@raspberrypi:~/adc/hat/c $ sudo ./main ADS1263 Demo /***********************************/ Current environment: Debian Please make JETSON !!!!!!!!!! Segmentation fault ...This is error I am getting

https://github.com/waveshareteam/High-Pricision_AD_HAT/blob/e8001efc40b998fefb0e06385e526fed9bf6bf37/c/lib/Config/DEV_Config.c#L221

It looks like there are some funny checks going on.
But it is the sysfs interface you probably want; there are funnys in the rpi5 chips

try #undef RPI #define JETSON , and delete the system checks. Hacky, experimental, but should get you there.

But do double check the pin outs here 
https://github.com/waveshareteam/High-Pricision_AD_HAT/blob/e8001efc40b998fefb0e06385e526fed9bf6bf37/c/lib/Config/DEV_Config.c#L243
